compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -446,12 +447,195 @@ impl<'a> Parser<'a> {
+                                                .map(|sp| sp.to(local.ty.as_ref().unwrap().span)),
+                                            has_initializer: !matches!(local.kind, LocalKind::Decl),
+                                        },
+                                    );
+
+                                    // don't suggest `let const { pat } = expr;`
+                                    return;
+                                }
+
+                                LocalKind::InitElse(_, _) => {}
+                            }
+                        }
+
+                        // help: use an arm guard `if val == expr`
+                        if let Some(arm) = &self.arm
+                            && !is_range_bound
+                        {
+                            let (ident, ident_span) = match self.field {
+                                Some(field) => {
+                                    (field.ident.to_string(), field.ident.span.to(expr_span))
+                                }
+                                None => ("val".to_owned(), expr_span),
+                            };
+
+                            match &arm.guard {
+                                None => {
+                                    err.subdiagnostic(
+                                        UnexpectedExpressionInPatternSugg::CreateGuard {
+                                            ident_span,
+                                            pat_hi: arm.pat.span.shrink_to_hi(),
+                                            ident,
+                                            expr: expr.clone(),
+                                        },
+                                    );
+                                }
+                                Some(guard) => {
+                                    err.subdiagnostic(
+                                        UnexpectedExpressionInPatternSugg::UpdateGuard {
+                                            ident_span,
+                                            guard_lo: guard.span.shrink_to_lo(),
+                                            guard_hi: guard.span.shrink_to_hi(),
+                                            ident,
+                                            expr: expr.clone(),
+                                        },
+                                    );
+                                }
+                            }
+                        }
+
+                        // help: extract the expr into a `const VAL: _ = expr`
+                        let ident = match self.field {
+                            Some(field) => field.ident.as_str().to_uppercase(),
+                            None => "VAL".to_owned(),
+                        };
+                        err.subdiagnostic(UnexpectedExpressionInPatternSugg::Const {
+                            stmt_lo: line_lo,
+                            ident_span: expr_span,
+                            expr,
```

Comment:
... but yeah, I don't know how we can avoid calling `span_to_snippet` for `UnexpectedExpressionInPatternSugg::Const`